### PR TITLE
[WIP] Completely rework shell scripts

### PR DESCRIPTION
**Pull request by c0d3z3r0**
### Motivation and Context

The shell scripts zfs-functions et al. have multiple problems:
- Current shell code is hard to read and debug because of too much unoptimized code
- Some bugs (like #9168, #9183, and I guess some more but I did not search in the issues, yet)
- Inconsistent code style
- Use of upper case variables (very confusing at least for me, as only env vars or globals should be upper case (isn't this (un)written convention?)
- Very much missing documentation which makes it hard for new devs to understand what SHOULD happen (which is not always what really happens :P)
- use of evi... eval() where it is not needed at all, e.g.: `eval export FSTAB_dev_$i="$fs"`
- duplicated functionality: `in_mtab` vs. `is_mounted`
- inconsistent use of `mount` and `cat/read /proc/self/mounts` (which both return the same results. The only differene is that mount does some formatting (insert `on` and `type`, put `options` in brackets, omitting `freq` and `pass`)
- Caching of fstab and mtab in env makes everything very, very complicated due to special character handling and (as far as I looked at the code, yet) is not needed at all. (MAYBE this is really needed and I did not get this far, yet. We'll see.) IMHO it really does not matter if we call grep multiple times but can completely drop any `eval`. The run time difference is negligible and should be in the range of some 10s or 100s of milliseconds.
- ... (more to follow)

This PR does a complete rework of all existing shell scripts to fix all those problems while keeping backwards compatibility.

Signed-off-by: Michael Niewöhner <foss@mniewoehner.de>

### How Has This Been Tested?

Currently only manual basic testing on function level;

### Types of changes

- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Performance enhancement (non-breaking change which improves efficiency)
- [x] Code cleanup (non-breaking change which makes code smaller or more readable)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [x] Documentation (a change to man pages or other documentation)

### Checklist:

- [x] My code follows the ZFS on Linux [code style requirements](https://github.com/zfsonlinux/zfs/blob/master/.github/CONTRIBUTING.md#coding-conventions).
- [ ] I have updated the documentation accordingly.
- [x] I have read the [**contributing** document](https://github.com/zfsonlinux/zfs/blob/master/.github/CONTRIBUTING.md).
- [ ] I have added [tests](https://github.com/zfsonlinux/zfs/tree/master/tests) to cover my changes.
- [ ] All new and existing tests passed.
- [x] All commit messages are properly formatted and contain [`Signed-off-by`](https://github.com/zfsonlinux/zfs/blob/master/.github/CONTRIBUTING.md#signed-off-by).

### CODE REVIEW

**Comment by gdevenyi** [comment]
Has shellcheck been used to check for issues?

**Comment by c0d3z3r0** [comment]
> Has shellcheck been used to check for issues?

not yet

**Comment by ahrens** [comment]
> This PR does a complete rework of all existing shell scripts

It looks like the commit here is changing two `.in` files (which are processed to generate shell scripts).  There are more shell scripts in `zed.d/`, `scripts/`, and many in the test suite.  Do the problems you described apply to those as well, and are you planning to change them?

**Comment by c0d3z3r0** [comment]
> > This PR does a complete rework of all existing shell scripts
> 
> It looks like the commit here is changing two `.in` files (which are processed to generate shell scripts). There are more shell scripts in `zed.d/`, `scripts/`, and many in the test suite. Do the problems you described apply to those as well, and are you planning to change them?

Yes, this is still WIP and I will push more changes over the next time, but it will take a bit because a) it's really much code and b) I don't want to break anything and c) want to keep backwards compatibility as good as possible

File: etc/init.d/zfs-functions.in [hunk 1]
```diff
@@ -132,33 +173,38 @@ zfs_action()
...
 {
-	local PIDFILE="$1";	shift
-	local DAEMON_BIN="$1";	shift
-	local DAEMON_ARGS="$*"
+	local pidfile="${1}";		shift
+	local daemon_bin="${1}";	shift
+	local daemon_args="$*"
 
-	if type start-stop-daemon > /dev/null 2>&1 ; then
+	if do_noout type start-stop-daemon
```
[the remark this answers is not in the document]
Comment (c0d3z3r0) [reply etc/init.d/zfs-functions.in]: IMHO we shouldn't quiet errors here because that would make debugging in case of errors very hard

File: etc/init.d/zfs-functions.in [hunk 2]
```diff
@@ -172,120 +218,141 @@ zfs_daemon_start()
...
 
-	if type start-stop-daemon > /dev/null 2>&1 ; then
+	if do_noout type start-stop-daemon; then
 		# LSB functions
-		start-stop-daemon --stop --quiet --retry=TERM/30/KILL/5 \
-		    --pidfile "$PIDFILE" --name "$DAEMON_NAME"
-		[ "$?" = 0 ] && rm -f "$PIDFILE"
+		start-stop-daemon --quiet --stop \
+		    --pidfile "${pidfile}" --name "${daemon_name}" \
+		    --retry=TERM/30/KILL/5
```
[the remark this answers is not in the document]
Comment (c0d3z3r0) [reply etc/init.d/zfs-functions.in]: will be deleted by the daemon(s), so wo do not need to do it manually

File: etc/init.d/zfs-functions.in [hunk 3]
```diff
@@ -172,120 +218,141 @@ zfs_daemon_start()
...
+# Test if a cmd is working
+# args:		cmd
+# output:	-
+# returns:	cmd return code
+test_zfs_cmd() {
+	local cmd=${1}
+
+	# Test if it works
+	# (will catch missing binary and missing/broken libs etc)
+	do_noout "${cmd}" -?
```
[the remark this answers is not in the document]
Comment (c0d3z3r0) [reply etc/init.d/zfs-functions.in]: no, we want to know if this fails and why, don't we?

File: etc/init.d/zfs-functions.in [hunk 4]
```diff
@@ -172,120 +218,141 @@ zfs_daemon_start()
...
+
+	return $?
```
[the remark this answers is not in the document]
Comment (c0d3z3r0) [reply etc/init.d/zfs-functions.in]: Indeed, and I've been struggling with that... personally I like simplicity, so this should be removed where not needed. On the other side, explicity is better than implicity, especially when reading code explicit returns make it a little bit easier to read.

Let's see what the others say

File: etc/init.d/zfs-functions.in [hunk 5]
```diff
@@ -172,120 +218,141 @@ zfs_daemon_start()
...
-
-	if [ ! -x "$ZFS" ]; then
-		return 2
-	else
-		# Test if it works (will catch missing/broken libs etc)
-		"$ZFS" -? > /dev/null 2>&1
-		return $?
-	fi
+	test_zfs_cmd "${ZPOOL}" || return 1
+	test_zfs_cmd "${ZFS}" || return 2
```
[the remark this answers is not in the document]
Comment (c0d3z3r0) [reply etc/init.d/zfs-functions.in]: because we would not know which one failed if one returned non-zero

File: etc/init.d/zfs-functions.in [hunk 6]
```diff
@@ -294,143 +361,206 @@ checksystem()
 	# Just make sure that /dev/zfs is created.
 	udev_trigger
 
-	if ! [ "$(uname -m)" = "x86_64" ]; then
-		echo "Warning: You're not running 64bit. Currently native zfs in";
-		echo "         Linux is only supported and tested on 64bit.";
+	if [ "$(uname -m)" != "x86_64" ]; then
+		echo "Warning:	You're not running 64bit. Currently native zfs in" >&2;
```
[the remark this answers is not in the document]
Comment (c0d3z3r0) [reply etc/init.d/zfs-functions.in]: good catch! regarding stderr... not sure, we could leave it stdout so it gets printed in any case

File: etc/init.d/zfs-functions.in [hunk 7]
```diff
@@ -294,143 +361,206 @@ checksystem()
...
+	[ $(echo "${lines}" | wc -l) -gt 1 ] && \
+	    echo 'Warning: multiple matches; only first one returned.' >&2
+
+	printf "${lines}" | head -1
+
 	return 0
 }
 
-# first parameter is a regular expression that filters mtab
-read_mtab()
```
[the remark this answers is not in the document]
Comment (c0d3z3r0) [reply etc/init.d/zfs-functions.in]: Oh guys, I'm repeating over and over again. This PR is still WIP. I will fix the others too, as promised but this will take some days ;-)

[the remark this answers is not in the document]
Comment (c0d3z3r0) [reply etc/init.d/zfs-functions.in]: heh what!? no, Github, I don't want to start a review on my own code... wtf...

File: etc/init.d/zfs-functions.in [hunk 8]
```diff
@@ -294,143 +361,206 @@ checksystem()
...
+# Get entry from fstab
+#
+# args:		filter (see filter_tabfile())
+# output:	matching mtab entry
+# returns:	boolean
+#
+get_fstab_entry()
```
[the remark this answers is not in the document]
Comment (c0d3z3r0) [reply etc/init.d/zfs-functions.in]: Yes, because I did not rewrite the users of the env vars. This will be used once that is done

File: etc/init.d/zfs-functions.in [hunk 9]
```diff
@@ -294,143 +361,206 @@ checksystem()
...
-	unset $(env | grep ^FSTAB_ | sed 's,=.*,,')
+	local tabfile="${1}";	shift
+	local compatfilter
 
-	i=0
-	while read -r fs mntpnt fstype opts; do
-		echo "$fs" | egrep -qE '^#|^$' && continue
-		echo "$mntpnt" | egrep -qE '^none|^swap' && continue
-		echo "$fstype" | egrep -qE '^swap' && continue
+	# backwards compatibility
```
[the remark this answers is not in the document]
Comment (c0d3z3r0) [reply etc/init.d/zfs-functions.in]: I wanted to provide backwards compatibility to maybe exisiting user/distributions scripts but we can decide to not support any of them and cross fingers we don't break their systems... not sure and this is not something I can decide. @behlendorf @ahrens what do you say about compat for external users?

File: etc/init.d/zfs-functions.in [hunk 10]
```diff
@@ -12,116 +12,157 @@
...
 	# Gentoo
 	. /etc/init.d/functions.sh
-elif [ -f /lib/lsb/init-functions ]; then
+elif [ -e /lib/lsb/init-functions ]; then
 	# LSB, Debian GNU/Linux and derivates
 	. /lib/lsb/init-functions
 fi
 
+# helpers
+do_noout() { $@ >/dev/null; }
```
[the remark this answers is not in the document]
Comment (c0d3z3r0) [reply etc/init.d/zfs-functions.in]: Yep, that's fixed already

File: etc/init.d/zfs-functions.in [hunk 11]
```diff
@@ -12,116 +12,157 @@
...
+# Source zfs configuration, overriding the defaults
+if [ -f @initconfdir@/zfs ]; then
+	. @initconfdir@/zfs
+fi
+
+# ----------------------------------------------------
+
 # Source function library
-if [ -f /etc/rc.d/init.d/functions ]; then
+if [ -e /etc/rc.d/init.d/functions ]; then
```
[the remark this answers is not in the document]
Comment (c0d3z3r0) [reply etc/init.d/zfs-functions.in]: uhm. wtf did I do there...? I wanted to change `-L` to `-f` as this matches both symlink and file...

File: etc/init.d/zfs-functions.in [hunk 12]
```diff
@@ -132,33 +173,38 @@ zfs_action()
...
-		daemon --pidfile "$PIDFILE" "$DAEMON_BIN" $DAEMON_ARGS
-		return $?
+		daemon --pidfile "${pidfile}" "${daemon_bin}" \
+		    ${daemon_args} || \
+		    return 3
+
 	else
 		# Unsupported
-		return 3
+		return -1
```
[the remark this answers is not in the document]
Comment (c0d3z3r0) [reply etc/init.d/zfs-functions.in]: context ;) I needed 3 for the return four lines above. Increasing to 4 didn't seem right as it is not related to 1-3 but a completely different error. Since -1 isn't posix this returns 255 now.

File: etc/init.d/zfs-functions.in [hunk 13]
```diff
@@ -294,143 +361,206 @@ checksystem()
...
+
+	if [ "${dev}${mntpt}${type}${opts}" = "" ]
+	then
+    		echo "Invalid filter: '$@'" >&2
+		return 1
 	fi
+
+	local lines="$( \
+	    egrep -av '^#' "${tabfile}" | \
+    	    egrep -ai "^${dev:-[^ ]+} ${mntpt:-[^ ]+} ${type:-[^ ]+} ${opts:-[^ ]+}" | \
```
[the remark this answers is not in the document]
Comment (c0d3z3r0) [reply etc/init.d/zfs-functions.in]: Well, I did catch \040 and \011 as the are named in the man page. I wasn't aware of other possible values. Does anyone really use that?

File: etc/init.d/zfs-functions.in [hunk 14]
```diff
@@ -294,143 +361,206 @@ checksystem()
...
+	    sed 's/ /\x00/g; s/\\\040/ /g; s/\\\011/\t/g'
+	    )"
+
+	[ $(echo "${lines}" | wc -l) -gt 1 ] && \
+	    echo 'Warning: multiple matches; only first one returned.' >&2
+
+	printf "${lines}" | head -1
```
[the remark this answers is not in the document]
Comment (c0d3z3r0) [reply etc/init.d/zfs-functions.in]: `\t` isn't allowed in fstab, is it?

**Comment by c0d3z3r0** [comment]
@chrisrd Thanks for looking at this!

> Huh. I could have sworn I posted a final comment previously. Oh well, it went something like this...
> 
> The processing of the mount tables definitely needs improvement, and more documentation is always appreciated.
> 
> I'm a bit equivocal about the style changes: I certainly don't dislike the new style as it matches my own when writing new scripts, however I don't find the existing style difficult, and bulk changes like this are a bit disruptive to the code base.

The problem I see is that there is no style. There are two or three, which makes it difficult and error-prone.

> 
> For the removal of mount table caching via environment variables (which is definitely flawed in the current code - _I have discovered a truly marvelous fix of this, which this margin is too narrow to contain_), can you characterise what the effect might be, e.g. for a system with, say, 30,000 live mounts?

Haha! I knew there must be a reason for that :D No, I absolutely can not... I will look at this an do some simulations with both codes so we can see the difference.

> 
> I wonder if `findmnt` could be used rather than hand rolling our own mount table interpreter, whichever way (cache or not) we decide to go. I guess that depends on whether `findmnt` is available on all supported platforms.


> 
> In the original and new code, `get_root_pool()` and `in_fstab()` are unused. It looks like removal of these functions means the only caller of `filter_tabfile()` would be `in_mtab()` (via `in_tabfile()`). That would allow sigificant simplification of the `filter_tabfile()` - removal of the argument parsing etc.

Here the same applies as for ... uhm, I forgot the function name... look in one of my comments above... they will (maybe) be needed later when I touch the other scripts. If we like to, we can drop any leftovers then.

File: etc/init.d/zfs-functions.in (hunk 1 above)
[the remark this answers is not in the document]
Comment (chrisrd) [reply etc/init.d/zfs-functions.in]: It's testing if `start-stop-daemon` or `daemon` is present, which depends on the distribution. If it's a `daemon` type distribution they don't need to see "start-stop-daemon: not found".

File: etc/init.d/zfs-functions.in (hunk 12 above)
[the remark this answers is not in the document]
Comment (chrisrd) [reply etc/init.d/zfs-functions.in]: I'm wondering if the return codes have meaning within the init system they're called from? E.g. "3" might *mean* "unsupported" and changing the return code might change the behavior of the init system.

File: etc/init.d/zfs-functions.in (hunk 2 above)
[the remark this answers is not in the document]
Comment (chrisrd) [reply etc/init.d/zfs-functions.in]: Are you sure? E.g.
```
$man start-stop-daemon
...
  -p, --pidfile pid-file
    Check whether a process has created the file pid-file. Note: using this 
    matching option alone might cause unintended processes to be acted on, if
    the old process terminated without being able to remove the pid-file.
...
```

File: etc/init.d/zfs-functions.in (hunk 3 above)
[the remark this answers is not in the document]
Comment (chrisrd) [reply etc/init.d/zfs-functions.in]: We're again testing the programs are present (and run), so maybe we don't need any other output than "ZFS not installed".

File: etc/init.d/zfs-functions.in (hunk 5 above)
[the remark this answers is not in the document]
Comment (chrisrd) [reply etc/init.d/zfs-functions.in]: In this case we don't look at the return code, other than pass/fail. When I see different return codes like that I think "ok, so what do those codes mean in the larger context?" and go looking. But in this case there's nothing there so I've wasted my time and energy.

File: etc/init.d/zfs-functions.in (hunk 6 above)
[the remark this answers is not in the document]
Comment (chrisrd) [reply etc/init.d/zfs-functions.in]: `stderr` certainly makes sense as they're error/warning messages, however in the larger context of the init system I'm once again wondering if `stdout`/`stderr` are treated differently, and which one we need here.

File: etc/init.d/zfs-functions.in (hunk 13 above)
[the remark this answers is not in the document]
Comment (chrisrd) [reply etc/init.d/zfs-functions.in]: It's a big world out there: _someone_ is sure to be using it! :smile:

Also, `/proc/self/mounts` encodes a `\` as `\134` so need to catch at least that as well. Here's some fun:
```
# mkdir '/tmp/foo' '/tmp/foo\t2\1 x\11'
# ls -ld '/tmp/foo' '/tmp/foo\t2\1 x\11'
drwxr-xr-x 2 root root 4096 Aug 21 12:19 /tmp/foo
drwxr-xr-x 2 root root 4096 Aug 21 12:19 /tmp/foo\t2\1 x\11
# echo '/tmp/foo /tmp/foo\t2\1341\040x\11 none bind,noauto' >> /etc/fstab
# mount '/tmp/foo\t2\1 x\11'
# grep foo /proc/self/mounts
/dev/mapper/ubuntu--vg-root /tmp/foo\134t2\1341\040x\13411 ext4 rw,relatime,errors=remount-ro,data=ordered 0 0
```

File: etc/init.d/zfs-functions.in (hunk 14 above)
[the remark this answers is not in the document]
Comment (chrisrd) [reply etc/init.d/zfs-functions.in]: Sure is - see above. Backslash only means something if followed by exactly 3 octal digits. The rest can be arbitrary bytes (but not null).

File: etc/init.d/zfs-functions.in (hunk 7 above)
[the remark this answers is not in the document]
Comment (chrisrd) [reply etc/init.d/zfs-functions.in]: The problem is incomplete changes will ensure the github auto tests will fail.

**Comment by gdevenyi** [comment]
Current state of shellcheck on this PR
```
In zfs-zed.in line 1:
#!@SHELL@
^-- SC1008: This shebang was unrecognized. Note that ShellCheck only handles sh/bash/dash/ksh.


In zfs-zed.in line 30:
. @sysconfdir@/zfs/zfs-functions
^-- SC1091: Not following: @sysconfdir@/zfs/zfs-functions was not specified as input (see shellcheck -x).


In zfs-zed.in line 35:
extra_started_commands="reload"
^-- SC2034: extra_started_commands appears unused. Verify use (or export if used externally).


In zfs-zed.in line 60:
	local pools RET
                    ^-- SC2034: RET appears unused. Verify use (or export if used externally).


In zfs-zed.in line 65:
	if [ "$?" -eq "0" ]
             ^-- SC2181: Check exit code directly with e.g. 'if mycmd;', not indirectly with $?.


In zfs-share.in line 1:
#!@SHELL@
^-- SC1008: This shebang was unrecognized. Note that ShellCheck only handles sh/bash/dash/ksh.


In zfs-share.in line 31:
. @sysconfdir@/zfs/zfs-functions
^-- SC1091: Not following: @sysconfdir@/zfs/zfs-functions was not specified as input (see shellcheck -x).


In zfs-mount.in line 1:
#!@SHELL@
^-- SC1008: This shebang was unrecognized. Note that ShellCheck only handles sh/bash/dash/ksh.


In zfs-mount.in line 32:
. @sysconfdir@/zfs/zfs-functions
^-- SC1091: Not following: @sysconfdir@/zfs/zfs-functions was not specified as input (see shellcheck -x).


In zfs-mount.in line 37:
	while read line; do
              ^-- SC2162: read without -r will mangle backslashes.


In zfs-mount.in line 38:
		set -- $line
                       ^-- SC2086: Double quote to prevent globbing and word splitting.


In zfs-mount.in line 68:
	local verbose overlay i mntpt val
                                      ^-- SC2034: val appears unused. Verify use (or export if used externally).


In zfs-import.in line 1:
#!@SHELL@
^-- SC1008: This shebang was unrecognized. Note that ShellCheck only handles sh/bash/dash/ksh.


In zfs-import.in line 35:
. @sysconfdir@/zfs/zfs-functions
^-- SC1091: Not following: @sysconfdir@/zfs/zfs-functions was not specified as input (see shellcheck -x).


In zfs-import.in line 59:
	do_noerr ${cmd} | egrep "pool:|^[a-zA-Z0-9]" | \
                 ^-- SC2086: Double quote to prevent globbing and word splitting.
                          ^-- SC2196: egrep is non-standard and deprecated. Use grep -E instead.


In zfs-import.in line 71:
	[ -n "$init" ] && rm -f /etc/dfs/sharetab
              ^-- SC2154: init is referenced but not assigned.


In zfs-import.in line 74:
	if [ -n "$USE_DISK_BY_ID" -a "$USE_DISK_BY_ID" != 'yes' ]
                                  ^-- SC2166: Prefer [ p ] && [ q ] as [ p -a q ] is not well defined.


In zfs-import.in line 102:
				npools=$(echo "$npools" | sed "s,$available_pools,,")
                                         ^-- SC2001: See if you can use ${variable//search/replace} instead.


In zfs-import.in line 147:
	if [ -n "$USE_DISK_BY_ID" -a -z "$ZPOOL_IMPORT_PATH" ]
                                  ^-- SC2166: Prefer [ p ] && [ q ] as [ p -a q ] is not well defined.


In zfs-import.in line 213:
		"$ZPOOL" import -N ${ZPOOL_IMPORT_OPTS} "$pool" 2> /dev/null
                                   ^-- SC2086: Double quote to prevent globbing and word splitting.


In zfs-import.in line 225:
		if [ "$r" -gt 0 -a -f "$ZPOOL_CACHE" ]
                                ^-- SC2166: Prefer [ p ] && [ q ] as [ p -a q ] is not well defined.


In zfs-import.in line 234:
			"$ZPOOL" import -c "$ZPOOL_CACHE" -N ${ZPOOL_IMPORT_OPTS} \
                                                             ^-- SC2086: Double quote to prevent globbing and word splitting.


In zfs-import.in line 248:
	[ -n "$already_imported" -a -z "$available_pools" ] && return 0
                                 ^-- SC2166: Prefer [ p ] && [ q ] as [ p -a q ] is not well defined.


In zfs-import.in line 255:
	check_boolean "${ZPOOL_IMPORT_ALL_VISIBLE}" && \
                                                    ^-- SC2015: Note that A && B || C is not if-then-else. C may run when A is true.


In zfs-functions.in line 1:
# This is a script with common functions etc used by zfs-import, zfs-mount,
^-- SC2148: Tips depend on target shell and yours is unknown. Add a shebang.


In zfs-functions.in line 29:
	. @initconfdir@/zfs
        ^-- SC1091: Not following: @initconfdir@/zfs was not specified as input (see shellcheck -x).


In zfs-functions.in line 37:
	. /etc/rc.d/init.d/functions
        ^-- SC1091: Not following: /etc/rc.d/init.d/functions was not specified as input (see shellcheck -x).


In zfs-functions.in line 40:
	. /etc/init.d/functions.sh
        ^-- SC1091: Not following: /etc/init.d/functions.sh was not specified as input (see shellcheck -x).


In zfs-functions.in line 43:
	. /lib/lsb/init-functions
        ^-- SC1091: Not following: /lib/lsb/init-functions was not specified as input (see shellcheck -x).


In zfs-functions.in line 80:
		[ "${1}" -eq 0 ] && success || failure
                                 ^-- SC2015: Note that A && B || C is not if-then-else. C may run when A is true.


In zfs-functions.in line 162:
	[ "${ret}" -eq 0 ] && zfs_log_end_msg ${ret} || \
                           ^-- SC2015: Note that A && B || C is not if-then-else. C may run when A is true.


In zfs-functions.in line 195:
		[ -f "${pidfile}" -a -d /run/sendsigs.omit.d ] && \
                                  ^-- SC2166: Prefer [ p ] && [ q ] as [ p -a q ] is not well defined.


In zfs-functions.in line 355:
		[ -n "${init}" ] && exit 3
                      ^-- SC2154: init is referenced but not assigned.


In zfs-functions.in line 383:
	    IFS="\0" read -r dev _;
                 ^-- SC1117: Backslash is literal in "\0". Prefer explicit escaping: "\\0".


In zfs-functions.in line 452:
	local dev="$(parsearg dev "$@")"
              ^-- SC2155: Declare and assign separately to avoid masking return values.


In zfs-functions.in line 453:
	local mntpt="$(parsearg mntpt "$@")"
              ^-- SC2155: Declare and assign separately to avoid masking return values.


In zfs-functions.in line 454:
	local type="$(parsearg type "$@")"
              ^-- SC2155: Declare and assign separately to avoid masking return values.


In zfs-functions.in line 455:
	local opts="$(parsearg opts "$@")"
              ^-- SC2155: Declare and assign separately to avoid masking return values.


In zfs-functions.in line 459:
		echo "Invalid filter: \'$*\'" >&2
                                      ^-- SC1117: Backslash is literal in "\'". Prefer explicit escaping: "\\'".
                                          ^-- SC1117: Backslash is literal in "\'". Prefer explicit escaping: "\\'".


In zfs-functions.in line 463:
	local lines="$( \
              ^-- SC2155: Declare and assign separately to avoid masking return values.


In zfs-functions.in line 469:
	[ $(printf "%s" "${lines}" | wc -l) -gt 1 ] && \
          ^-- SC2046: Quote this to prevent word splitting.

```

**Comment by c0d3z3r0** [comment]
> Current state of shellcheck on this PR
> 
> ```
> In zfs-zed.in line 1:

Can you wait with this until I have more stuff ready, please? This is WIP (work in progress).... I just finished `zfs-functions` but there is work to be done as you can see at the conversation above.

Fyi I ran shellcheck yesterday and already applied some stuff to `zfs-functions` in my last push.

> In zfs-functions.in line 1:
> # This is a script with common functions etc used by zfs-import, zfs-mount,
> ^-- SC2148: Tips depend on target shell and yours is unknown. Add a shebang.

No. This script will NEVER run on it's own, so we do NOT wan't a shebang.

> 
> In zfs-functions.in line 29:
> 	. @initconfdir@/zfs
>         ^-- SC1091: Not following: @initconfdir@/zfs was not specified as input (see shellcheck -x).
> 
> 
> In zfs-functions.in line 37:
> 	. /etc/rc.d/init.d/functions
>         ^-- SC1091: Not following: /etc/rc.d/init.d/functions was not specified as input (see shellcheck -x).
> 
> 
> In zfs-functions.in line 40:
> 	. /etc/init.d/functions.sh
>         ^-- SC1091: Not following: /etc/init.d/functions.sh was not specified as input (see shellcheck -x).
> 
> 
> In zfs-functions.in line 43:
> 	. /lib/lsb/init-functions
>         ^-- SC1091: Not following: /lib/lsb/init-functions was not specified as input (see shellcheck -x).

Shellcheck default... scripts will be check one-for-one.

> In zfs-functions.in line 80:
> 		[ "${1}" -eq 0 ] && success || failure
>                                  ^-- SC2015: Note that A && B || C is not if-then-else. C may run when A is true.
> 
> 
> In zfs-functions.in line 162:
> 	[ "${ret}" -eq 0 ] && zfs_log_end_msg ${ret} || \
>                            ^-- SC2015: Note that A && B || C is not if-then-else. C may run when A is true.

Does not matter here.

> 
> In zfs-functions.in line 195:
> 		[ -f "${pidfile}" -a -d /run/sendsigs.omit.d ] && \
>                                   ^-- SC2166: Prefer [ p ] && [ q ] as [ p -a q ] is not well defined.

Not changed. Worked before and should be ok (I never new `-a` was not well defined anywhere. This may apply to vanilla POSIX only.

> 
> In zfs-functions.in line 355:
> 		[ -n "${init}" ] && exit 3
>                       ^-- SC2154: init is referenced but not assigned.

Not sure but I guess init is defined elsewhere  but it should be upercase then. Have to check.

> 
> In zfs-functions.in line 383:
> 	    IFS="\0" read -r dev _;
>                  ^-- SC1117: Backslash is literal in "\0". Prefer explicit escaping: "\\0".

Huh? Yes, of course. This shall be \0.

> 
> In zfs-functions.in line 452:
> 	local dev="$(parsearg dev "$@")"
>               ^-- SC2155: Declare and assign separately to avoid masking return values.
> 
> 
> In zfs-functions.in line 453:
> 	local mntpt="$(parsearg mntpt "$@")"
>               ^-- SC2155: Declare and assign separately to avoid masking return values.
> 
> 
> In zfs-functions.in line 454:
> 	local type="$(parsearg type "$@")"
>               ^-- SC2155: Declare and assign separately to avoid masking return values.
> 
> 
> In zfs-functions.in line 455:
> 	local opts="$(parsearg opts "$@")"
>               ^-- SC2155: Declare and assign separately to avoid masking return values.
> 
> In zfs-functions.in line 463:
> 	local lines="$( \
>               ^-- SC2155: Declare and assign separately to avoid masking return values.

I guess shellcheck is confused because of using opts as string here.

> 
> In zfs-functions.in line 459:
> 		echo "Invalid filter: \'$*\'" >&2
>                                       ^-- SC1117: Backslash is literal in "\'". Prefer explicit escaping: "\\'".
>                                           ^-- SC1117: Backslash is literal in "\'". Prefer explicit escaping: "\\'".

Nope. We do mean that literal.

Same as 
> In zfs-functions.in line 469:
> 	[ $(printf "%s" "${lines}" | wc -l) -gt 1 ] && \
>           ^-- SC2046: Quote this to prevent word splitting.
> ```

Does not matter here

**Comment by gdevenyi** [comment]
@c0d3z3r0 sorry, I updated my post to change how shellcheck is run, to enforce POSIXness and to allow it to import the other function files. I did grab your latest push as of when the check was run.

The shell this code is run against is /bin/sh which I understand could be dash or sh, but I think we won't know which, so we should stick to POSIX? Especially if we are integrating the BSDs into the repo eventually.

**Comment by c0d3z3r0** [comment]
> @c0d3z3r0 sorry, I updated my post to change how shellcheck is run, to enforce POSIXness and to allow it to import the other function files. I did grab your latest push as of when the check was run.
> 
> The shell this code is run against is /bin/sh which I understand could be dash or sh, but I think we won't know which, so we should stick to POSIX? Especially if we are integrating the BSDs into the repo eventually.

You can select the shell for shellcheck with `-s`. I guess we should check them all, even bash since some distros (arch linux for example) use bash in posix compatibility mode (calling it as "sh").
I am not sure if `ksh` is used anywhere is `/bin/sh`.

TL;DR yes we should stick to POSIX (but even POSIX compatible shells are not always implemented that strict in the shells; see the `-a` "problem" above).

**Comment by gdevenyi** [comment]
Great, the current status of my original comment includes an explicit "-s sh" in the shellcheck call, so it is enforcing strict POSIX.

File: etc/init.d/zfs-functions.in (hunk 1 above)
[the remark this answers is not in the document]
Comment (c0d3z3r0) [reply etc/init.d/zfs-functions.in]: Point for you, will fix that

File: etc/init.d/zfs-functions.in (hunk 2 above)
[the remark this answers is not in the document]
Comment (c0d3z3r0) [reply etc/init.d/zfs-functions.in]: Not start-stop-daemon but the daemon itself creates and deleetes the PID file.
Just look at zed for example :-) I checked that

File: etc/init.d/zfs-functions.in (hunk 13 above)
[the remark this answers is not in the document]
Comment (c0d3z3r0) [reply etc/init.d/zfs-functions.in]: Holy.... \o/ ok... maybe findmnt will fix this all for us...

File: etc/init.d/zfs-functions.in (hunk 7 above)
[the remark this answers is not in the document]
Comment (c0d3z3r0) [reply etc/init.d/zfs-functions.in]: Will be as soon as I have time

File: etc/init.d/zfs-functions.in (hunk 14 above)
[the remark this answers is not in the document]
Comment (c0d3z3r0) [reply etc/init.d/zfs-functions.in]: Meh.

**Comment by Ornias1993** [comment]
@gdevenyi  would it be an idea to add shellcheck to the general test suit anyway?
In that case we could makeup an issue for it.

**Comment by gdevenyi** [comment]
@Ornias1993 if the rework makes it in, adding shellcheck to the testsuite could prevent bugs/inconsistencies creeping back in

**Comment by Ornias1993** [comment]
@gdevenyi Wouldn't shellcheck be a good thing in the test suite, regardless if this PR is merged or another form of rework is done in the future?

**Comment by gdevenyi** [comment]
@Ornias1993 yes, but its likely to throw a bunch of errors if added now, since the scripts haven't been cleaned up. Typically I've seen PRs for such checks held back until the scripts pass and then merged in together.
